### PR TITLE
Memo transactions

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Add support for memo transactions

## Changes

Added support for memo transactions (simple, encrypted and with schedule).

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: doc/ins_transfer.md [hunk 1]
```diff
@@ -9,3 +9,18 @@ A transaction to transfer GTU from one account to another.
...
+
+## Protocol description
+
+* Multiple commands
+
+INS | P1 | P2 | CDATA | Comment |
+|----|--------|-----|-------------|----|
+| `0x02` | `0x01` | `0x00` | `path_length path[uint32]x[8] account_transaction_header[60 bytes] transaction_kind[uint8] recipient_address[32 bytes] memo_length[uint16]` | The amount is in µGTU. The recipient address has to be base58. |
+| `0x02` | `0x02` | `0x00` | `memo[1...255 bytes]` | The memo is assumed to be CBOR encoded. |
+| `0x02` | `0x03` | `0x00` | `amount[uint64]` | The memo is assumed to be CBOR encoded. |
```
Comment (orhoj) [inline doc/ins_transfer.md:26]: The comment does not fit here.

File: src/globals.h [hunk 2]
```diff
@@ -109,6 +114,15 @@ typedef struct {
...
 
+typedef struct {
+    union {
+        signTransferContext_t signTransferContext;
+        signEncryptedAmountToTransfer_t signEncryptedAmountToTransfer;
+        signTransferWithScheduleContext_t signTransferWithScheduleContext;
+    };
+    memoContext_t memoContext;
+
+} withMemo_t;
```
Comment (orhoj) [inline src/globals.h:125]: I would call this `transferWithMemo_t` instead to make it less generic.

File: src/memo.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    if (ctx->memoLength < 0) {
+        THROW(ERROR_INVALID_STATE);
+    } else {
+        sendSuccessNoIdle();   // Request more data from the computer.
+    }
+}
+
+void readMemoInitial(uint8_t *cdata, uint8_t dataLength) {
+    uint8_t header = cdata[0];
+    cdata+=1;
```
Comment (orhoj) [inline src/memo.c:34]: `cdata += 1;`

File: src/memo.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#ifndef _MEMO_H_
+#define _MEMO_H_
+
+void readMemoInitial(uint8_t *cdata, uint8_t dataLength);
```
Comment (orhoj) [inline src/memo.h:4]: Please add documentation. In particular it should state that it expects the memo to be CBOR encoded, and mention any other restrictions as well.

File: src/memo.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    uint8_t shortCount = header & 0x1f;
+
+    // Calculate length of cbor payload
+    // sizeLength: number of bytes (beside the header) used to indicate the payload length
+    uint8_t sizeLength = 0;
+    // length: payload byte size.
+    uint64_t length = 0;
+
+    if (shortCount < 24) {
+        // shortCount is the length, no extra bytes is used.
```
Comment (orhoj) [inline src/memo.c:47]: no extra bytes are used

File: src/memo.c [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        sizeLength = 0;
+        length = shortCount;
+    } else if (shortCount == 24) {
+        length = cdata[0];
+        sizeLength= 1;
```
Comment (orhoj) [inline src/memo.c:52]: `sizeLength = 1;`

File: src/memo.c [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    } else if (shortCount == 25) {
+        length = U2BE(cdata,0);
```
Comment (orhoj) [inline src/memo.c:54]: length = U2BE(cdata, 0);

File: src/memo.c [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    } else if (shortCount == 26) {
+        length = U4BE(cdata,0);
+        sizeLength = 4;
+    } else if (shortCount == 27) {
+        length = U8BE(cdata,0);
+        sizeLength = 8;
+    } else if (shortCount == 31) {
+        THROW(ERROR_UNSUPPORTED_CBOR);
+    } else {
+        THROW(ERROR_INVALID_STATE);
```
Comment (orhoj) [inline src/memo.c:65]: This error is unrelated to the state. I think `ERROR_INVALID_PARAM` is more fitting as it is invalid input that is not based on an earlier promise by the caller.

File: src/memo.c [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    cdata += sizeLength;
+    ctx->memoLength -= sizeLength;
+
+    switch (ctx->majorType) {
+    case 0:
```
Comment (orhoj) [inline src/memo.c:71]: Please indent the cases.

File: src/memo.c [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        // non-negative integer
+        bin2dec(ctx->memo, length);
+        if (ctx->memoLength != 0) {
+            THROW(ERROR_INVALID_STATE);
+        }
+        break;
+    case 1:
+        // negative integer
+        bin2dec(ctx->memo, 1 - length);
```
Comment (orhoj) [inline src/memo.c:80]: `bin2dec` expects an unsigned integer. Also the wikipedia article on CBOR says it should be `-1-length`.

File: src/memo.c [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        THROW(ERROR_INVALID_STATE);
+    }
+    cdata += sizeLength;
+    ctx->memoLength -= sizeLength;
+
+    switch (ctx->majorType) {
+    case 0:
+        // non-negative integer
+        bin2dec(ctx->memo, length);
+        if (ctx->memoLength != 0) {
```
Comment (orhoj) [inline src/memo.c:74]: This check is duplicated in three places, but as I see it could be done right after:
```
ctx->memoLength -= sizeLength;
```

File: src/memo.c [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        readMemoContent(cdata, dataLength - 1 - sizeLength);
+        break;
+    default:
+        THROW(ERROR_UNSUPPORTED_CBOR);
+    }
+}
+
+void readMemoContent(uint8_t *cdata, uint8_t dataLength) {
+    ctx->memoLength -= dataLength;
+    switch (ctx->majorType) {
```
Comment (orhoj) [inline src/memo.c:99]: As we only handle one case with this method, you can remove the switch case.

File: src/memo.c [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    case 3:
+        memmove(ctx->memo, cdata, dataLength);
+
+        if (dataLength < 255) {
+            memmove(ctx->memo + dataLength, "\0", 1);
```
Comment (orhoj) [inline src/memo.c:104]: I would expect that you have to terminate the string even if it is 255 bytes.

File: src/memo.c [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            THROW(ERROR_INVALID_STATE);
+        }
+        readMemoContent(cdata, dataLength - 1 - sizeLength);
+        break;
+    default:
+        THROW(ERROR_UNSUPPORTED_CBOR);
+    }
+}
+
+void readMemoContent(uint8_t *cdata, uint8_t dataLength) {
```
Comment (orhoj) [inline src/memo.c:97]: Please rename `dataLength` as it leads you believe that it is same `dataLength` that we get from the APDU size, which isn't necessarily the case.

File: src/signTransfer.c [hunk 15]
```diff
@@ -31,40 +34,118 @@ UX_FLOW(ux_sign_flow,
...
+        }
+        // Tell the main process to wait for a button press.
+        *flags |= IO_ASYNCH_REPLY;
+    } else if (p1 == P1_MEMO && ctx->state == TX_TRANSFER_MEMO_INITIAL) {
+        cx_hash((cx_hash_t *) &tx_state->hash, 0, cdata, dataLength, NULL, 0);
+
+        // Read initial part of memo and then display it:
+        readMemoInitial(cdata, dataLength);
+        ctx->state = TX_TRANSFER_MEMO;
+        ux_flow_init(0, ux_sign_transfer_memo, NULL);
```
Comment (orhoj) [inline src/signTransfer.c:129]: Wait until the full memo has been processed before displaying it.

**Comment by orhoj** [comment]
I think the signing methods are harder to read now that they handle multiple transaction types with different states. I think it can be improved by separating them into separate `INS` values and each having its own entry point. With them interleaved it is harder to follow.

File: src/memo.c (hunk 14 above)
Comment (shjortConcordium) [reply src/memo.c]: Renamed to contentLength

**Comment by shjortConcordium** [comment]
> I think the signing methods are harder to read now that they handle multiple transaction types with different states. I think it can be improved by separating them into separate `INS` values and each having its own entry point. With them interleaved it is harder to follow.

I have split them up, trying to reuse code between memo and non-memo versions. There is the `handleHeaderAndToAddress...` functions, which I feel should be possible to merge, but I don't see a nice way to do it.

File: src/memo.h (hunk 4 above)
Comment (orhoj) [reply src/memo.h:4]: The main documentation should be in the header file and not in the implementation file.

File: src/memo.c [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+            // non-negative integer
+            bin2dec(ctx->memo, length);
+            if (ctx->memoLength != 0) {
+                THROW(ERROR_INVALID_STATE);
+            }
+            break;
+        case 1:
+            // negative integer
+            memmove(ctx->memo, "-", 1);
+            bin2dec(ctx->memo + 1, 1 + length);
```
Comment (orhoj) [inline src/memo.c:87]: 1 + length can overflow if `length = maxUint64`.

**Comment by orhoj** [comment]
The Ledger accepts memos that are longer than what the blockchain accepts.

File: src/signTransfer.c [hunk 17]
```diff
@@ -68,3 +104,63 @@ void handleSignTransfer(uint8_t *cdata, volatile unsigned int *flags) {
...
+        // Read initial part of memo and then display it:
+        readMemoInitial(cdata, dataLength);
+
+        if (memo_ctx->memoLength == 0) {
+            finishMemo(flags);
+        } else {
+            ctx->state = TX_TRANSFER_MEMO;
+            sendSuccessNoIdle();
+        }
+    } else if (p1 == P1_MEMO && ctx->state == TX_TRANSFER_MEMO) {
```
Comment (orhoj) [inline src/signTransfer.c:142]: You have two places that check for `p1 == P1_MEMO`. Usually we have had separate `p1` values for the different steps.

File: src/memo.h (hunk 4 above)
Comment (shjortConcordium) [reply src/memo.h:4]: Moved it.

File: src/memo.c (hunk 16 above)
Comment (shjortConcordium) [reply src/memo.c]: Added special handling for this case.

**Comment by shjortConcordium** [comment]
> > > I think the signing methods are harder to read now that they handle multiple transaction types with different states. I think it can be improved by separating them into separate `INS` values and each having its own entry point. With them interleaved it is harder to follow.
> > 
> > 
> > I have split them up, trying to reuse code between memo and non-memo versions.
> > I feel that I should be able to merge the `handleHeaderAndToAddress...` functions, but I don't see a nice way to do it.
> 
> I tried to refactor those methods into a shared one and pushed it.

Looks good, and the transactions still work.

File: src/signTransfer.c (hunk 17 above)
Comment (shjortConcordium) [reply src/signTransfer.c:112]: Yes, It is because we want to handle the initial part of the memo differently, but we want the other side to be able to just chunk it naively.

**Review by orhoj — APPROVED** [review]
